File: app/jobs/duplicate_detection_job.rb
class DuplicateDetectionJob < ApplicationJob
  queue_as :default

  retry_on StandardError, wait: ->(executions) { (executions ** 4) + rand(30) }, attempts: 3

  # Don't retry if contact was deleted
  discard_on ActiveRecord::RecordNotFound do |job, exception|
    Rails.logger.error("[DuplicateDetectionJob] Contact not found: #{exception.message}")
  end

  # Note: This job expects a contact_id (Integer), not a Contact object
  # This follows Sidekiq best practice of passing IDs to avoid serialization issues
  def perform(contact_id)
    contact = Contact.find(contact_id)

    # Skip if already marked as duplicate
    return if contact.is_duplicate?

    credentials = TwilioCredential.current
    return if credentials && !credentials.enable_duplicate_detection

    Rails.logger.info("Checking for duplicates: contact #{contact.id}")

    # Find potential duplicates
    duplicates = DuplicateDetectionService.find_duplicates(contact)

    if duplicates.any?
      Rails.logger.info("Found #{duplicates.length} potential duplicates for contact #{contact.id}")

      # Store duplicate check timestamp
      contact.update!(duplicate_checked_at: Time.current)

      # Auto-merge if enabled and high confidence
      if credentials&.auto_merge_duplicates
        duplicates.each do |dup|
          if dup[:confidence] >= 95
            Rails.logger.info("Auto-merging contact #{dup[:contact].id} into #{contact.id} (confidence: #{dup[:confidence]}%)")
            DuplicateDetectionService.merge(contact, dup[:contact])
          end
        end
      end
    else
      Rails.logger.info("No duplicates found for contact #{contact.id}")
      contact.update!(duplicate_checked_at: Time.current)
    end

  rescue StandardError => e
    Rails.logger.error("Duplicate detection error for contact #{contact.id}: #{e.message}")
    raise # Re-raise to trigger retry logic
  end
end

File: laravel-app/app/Jobs/DuplicateDetectionJob.php
<?php

namespace App\Jobs;

use App\Models\Contact;
use App\Models\TwilioCredential;
use App\Services\DuplicateDetectionService;
use Illuminate\Bus\Queueable;
use Illuminate\Contracts\Queue\ShouldQueue;
use Illuminate\Foundation\Bus\Dispatchable;
use Illuminate\Queue\InteractsWithQueue;
use Illuminate\Queue\SerializesModels;
use Illuminate\Support\Facades\Log;
use Exception;

class DuplicateDetectionJob implements ShouldQueue
{
    use Dispatchable, InteractsWithQueue, Queueable, SerializesModels;

    public $contactId;

    /**
     * The number of times the job may be attempted.
     *
     * @var int
     */
    public $tries = 3;

    /**
     * The number of seconds to wait before retrying.
     *
     * @var int
     */
    public $backoff = [10, 30, 90];

    /**
     * The maximum number of seconds the job can run before timing out.
     *
     * @var int
     */
    public $timeout = 60;

    /**
     * Create a new job instance.
     *
     * @param int $contactId
     */
    public function __construct(int $contactId)
    {
        $this->contactId = $contactId;
        $this->onQueue('default');
    }

    /**
     * Execute the job.
     *
     * @return void
     */
    public function handle()
    {
        try {
            $contact = Contact::findOrFail($this->contactId);

            // Skip if already marked as duplicate
            if ($contact->is_duplicate) {
                return;
            }

            $credentials = TwilioCredential::current();
            if ($credentials && !$credentials->enable_duplicate_detection) {
                return;
            }

            Log::info("Checking for duplicates: contact {$contact->id}");

            // Find potential duplicates
            $service = new DuplicateDetectionService();
            $duplicates = $service->findDuplicates($contact);

            if (count($duplicates) > 0) {
                Log::info("Found " . count($duplicates) . " potential duplicates for contact {$contact->id}");

                // Store duplicate check timestamp
                $contact->update(['duplicate_checked_at' => now()]);

                // Auto-merge if enabled and high confidence
                if ($credentials && $credentials->auto_merge_duplicates) {
                    foreach ($duplicates as $dup) {
                        if ($dup['confidence'] >= 95) {
                            Log::info("Auto-merging contact {$dup['contact']->id} into {$contact->id} (confidence: {$dup['confidence']}%)");
                            $service->merge($contact, $dup['contact']);
                        }
                    }
                }
            } else {
                Log::info("No duplicates found for contact {$contact->id}");
                $contact->update(['duplicate_checked_at' => now()]);
            }

        } catch (Exception $e) {
            Log::error("Duplicate detection error for contact {$this->contactId}: {$e->getMessage()}");
            throw $e; // Re-raise to trigger retry logic
        }
    }

    /**
     * Handle a job failure.
     *
     * @param  \Throwable  $exception
     * @return void
     */
    public function failed(\Throwable $exception)
    {
        Log::error("[DuplicateDetectionJob] Failed for contact {$this->contactId}: {$exception->getMessage()}");
    }
}
